### PR TITLE
Improve WebUtility.HtmlEncode / UrlEncode performance

### PR DESCRIPTION
**Pull request by stephentoub**
- For HtmlEncode, vectorize IndexOfHtmlEncodingChars. Using SearchValues, we can efficiently search for the first ASCII encoding char or the first non-ASCII char, and only then fall back to a scalar loop.
- For HtmlEncode, reduce branching by using a more efficient check to determine whether the ASCII characters need to be encoded.
- For UrlEncode, rather than UTF8-encoding into a new byte[], %-encoding in-place in that, and then creating a string from that, we can use string.Create and just do all the encoding in that buffer.
- For UrlEncode, use SearchValues to vectorize the search for the first non-safe char. Also move the check for ' ' to be inside the if for non-safe char.
- For UrlEncode, use SearchValues to optimize the check for whether an individual character is part of the set (via Contains).
- Simplify IsUrlSafeChar. Rather than multiple checks, one of which is a bitmap, just have a bitmap.
- Remove some leading IsNullOrEmpty checks. Null/empty inputs should be rare, and they're now handled implicitly as part of the subsequent loops.

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Net;

BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args);

[MemoryDiagnoser]
public class Tests
{
    [Benchmark]
    [ArgumentsSource(nameof(Inputs))]
    public string HtmlEncode(string input) => WebUtility.HtmlEncode(input);

    [Benchmark]
    [ArgumentsSource(nameof(Inputs))]
    public string UrlEncode(string input) => WebUtility.UrlEncode(input);

    public IEnumerable<object[]> Inputs() =>
    [
        ["this-is-a-very-long-filename-for-an-image-that-should-not-need-encoding.jpg"],
        ["short_name.txt"],
        ["https://www.example.com"],
        ["<test>hello, world</test>"],
        ["לילה טוב"],
        ["""
         How much wood could a woodchuck chuck
         If a woodchuck could chuck wood?
         A woodchuck would chuck as much wood
         As much wood as a woodchuck could chuck,
         If a woodchuck could chuck wood.
         """]
    ];
}
```

| Method     | Toolchain         | input                | Mean       | Ratio | Allocated | Alloc Ratio |
|----------- |------------------ |--------------------- |-----------:|------:|----------:|------------:|
| HtmlEncode | \main\corerun.exe | <test(...)test> [25] |  81.576 ns |  1.00 |      96 B |        1.00 |
| HtmlEncode | \pr\corerun.exe   | <test(...)test> [25] |  80.366 ns |  0.97 |      96 B |        1.00 |
|            |                   |                      |            |       |           |             |
| UrlEncode  | \main\corerun.exe | <test(...)test> [25] | 118.515 ns |  1.00 |     160 B |        1.00 |
| UrlEncode  | \pr\corerun.exe   | <test(...)test> [25] | 103.039 ns |  0.87 |      96 B |        0.60 |
|            |                   |                      |            |       |           |             |
| HtmlEncode | \main\corerun.exe | How (...)ood. [185]  | 164.359 ns |  1.00 |         - |          NA |
| HtmlEncode | \pr\corerun.exe   | How (...)ood. [185]  |  10.142 ns |  0.06 |         - |          NA |
|            |                   |                      |            |       |           |             |
| UrlEncode  | \main\corerun.exe | How (...)ood. [185]  | 537.945 ns |  1.00 |     664 B |        1.00 |
| UrlEncode  | \pr\corerun.exe   | How (...)ood. [185]  | 466.832 ns |  0.86 |     432 B |        0.65 |
|            |                   |                      |            |       |           |             |
| HtmlEncode | \main\corerun.exe | https(...)e.com [23] |  21.442 ns |  1.00 |         - |          NA |
| HtmlEncode | \pr\corerun.exe   | https(...)e.com [23] |   4.732 ns |  0.22 |         - |          NA |
|            |                   |                      |            |       |           |             |
| UrlEncode  | \main\corerun.exe | https(...)e.com [23] | 103.053 ns |  1.00 |     136 B |        1.00 |
| UrlEncode  | \pr\corerun.exe   | https(...)e.com [23] |  76.013 ns |  0.73 |      80 B |        0.59 |
|            |                   |                      |            |       |           |             |
| HtmlEncode | \main\corerun.exe | short_name.txt       |  13.068 ns |  1.00 |         - |          NA |
| HtmlEncode | \pr\corerun.exe   | short_name.txt       |   4.876 ns |  0.38 |         - |          NA |
|            |                   |                      |            |       |           |             |
| UrlEncode  | \main\corerun.exe | short_name.txt       |  14.324 ns |  1.00 |         - |          NA |
| UrlEncode  | \pr\corerun.exe   | short_name.txt       |   3.600 ns |  0.26 |         - |          NA |
|            |                   |                      |            |       |           |             |
| HtmlEncode | \main\corerun.exe | this-(...)g.jpg [75] |  62.800 ns |  1.00 |         - |          NA |
| HtmlEncode | \pr\corerun.exe   | this-(...)g.jpg [75] |   6.975 ns |  0.11 |         - |          NA |
|            |                   |                      |            |       |           |             |
| UrlEncode  | \main\corerun.exe | this-(...)g.jpg [75] |  70.709 ns |  1.00 |         - |          NA |
| UrlEncode  | \pr\corerun.exe   | this-(...)g.jpg [75] |   5.611 ns |  0.08 |         - |          NA |
|            |                   |                      |            |       |           |             |
| HtmlEncode | \main\corerun.exe | לילה טוב             |   8.436 ns |  1.00 |         - |          NA |
| HtmlEncode | \pr\corerun.exe   | לילה טוב             |  11.427 ns |  1.36 |         - |          NA |
|            |                   |                      |            |       |           |             |
| UrlEncode  | \main\corerun.exe | לילה טוב             | 100.852 ns |  1.00 |     184 B |        1.00 |
| UrlEncode  | \pr\corerun.exe   | לילה טוב             |  74.544 ns |  0.74 |     112 B |        0.61 |

### CODE REVIEW

**Review by MihaZupan — APPROVED** [review]
Always love to see SearchValues ratios like these

| Method     | Toolchain         | input                | Mean       | Ratio | Allocated | Alloc Ratio |
|----------- |------------------ |--------------------- |-----------:|------:|----------:|------------:|
| HtmlEncode | \main\corerun.exe | How (...)ood. [185]  | 164.359 ns |  1.00 |         - |          NA |
| HtmlEncode | \pr\corerun.exe   | How (...)ood. [185]  |  10.142 ns |  0.06 |         - |          NA |

File: src/libraries/System.Private.CoreLib/src/System/Net/WebUtility.cs [hunk 1]
```diff
@@ -291,28 +271,25 @@ private static void HtmlDecode(ReadOnlySpan<char> input, ref ValueStringBuilder
...
-            for (int i = 0; i < input.Length; i++)
+            for (int i = input.IndexOfAnyExcept(s_htmlAsciiNonEncodingChars); (uint)i < (uint)input.Length; i++)
             {
                 char ch = input[i];
                 if (ch <= '>')
                 {
-                    switch (ch)
+                    // Matches any of <>&'"
+                    uint charMinusLowUInt32 = (ushort)(ch - '"');
+                    if ((int)((0x8C000028U << (short)charMinusLowUInt32) & (charMinusLowUInt32 - 32)) < 0)
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Net/WebUtility.cs:287]: It's similar but it still has an extra branch. Not worth it?

[the remark this answers is not in the document]
Comment (MihaZupan) [reply src/libraries/System.Private.CoreLib/src/System/Net/WebUtility.cs:287]: > extra branch

I'm guessing that's what @EgorBo was referring to here: https://github.com/dotnet/runtime/pull/87656#issuecomment-1600812074.
Feels like something we should eventually decide on which is better and have the JIT emit that to keep callers more readable.

It's fine as-is in this PR.

[the remark this answers is not in the document]
Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Net/WebUtility.cs:287]: It looks like jit's recognition needs some extra work to optimize it, but it already emits a bitmap test for `if (ch == '<' || ch == '>' || ch == '"' || ch == '\'' || ch == '&')` just with extra branches so it's up to you.

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Net/WebUtility.cs:287]: I'll change it, if for no other reason than easier-to-maintain code. But it'd be good if the JIT could be improved here, in particular for when using the more concise pattern matching syntax, e.g. the JIT optimizes:
```C#
if (ch == '<' || ch == '>' || ch == '"' || ch == '\'' || ch == '&')
```
but not
```C#
if (c is '<' or '>' or '"' or '\'' or '&')
```
The latter is considered newer / cleaner C# and Visual Studio refactorings push you towards using it, so it's unfortunate that if someone does such a change they then regress.

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Net;

BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args);

[MemoryDiagnoser]
public class Tests
{
    private char[] _chars;

    [Params(1, 2, 3)]
    public int Mode { get; set; }

    [GlobalSetup]
    public void Setup()
    {
        var r = new Random(42);
        switch (Mode)
        {
            case 1:
                _chars = r.GetItems<char>("<>\"'&", 1000);
                break;
            case 2:
                _chars = r.GetItems<char>("0123456789abcdefghijklmnopqrstuvwxyz", 1000);
                break;
            case 3:
                _chars = r.GetItems<char>("<>\"'&0123456789abcdefghijklmnopqrstuvwxyz", 1000);
                break;
        }
    }

    [Benchmark(Baseline = true)]
    public int CountManual()
    {
        int count = 0;
        foreach (char c in _chars)
        {
            uint charMinusLowUInt32 = (ushort)(c - '"');
            if ((int)((0x8C000028U << (short)charMinusLowUInt32) & (charMinusLowUInt32 - 32)) < 0)
            {
                count++;
            }
        }
        return count;
    }

    [Benchmark]
    public int CountJit()
    {
        int count = 0;
        foreach (char c in _chars)
        {
            if (c == '<' || c == '>' || c == '"' || c == '\'' || c == '&')
            {
                count++;
            }
        }
        return count;
    }

    [Benchmark]
    public int CountPattern()
    {
        int count = 0;
        foreach (char c in _chars)
        {
            if (c is '<' or '>' or '"' or '\'' or '&')
            {
                count++;
            }
        }
        return count;
    }
}
```

| Method       | Mode | Mean       | Ratio |
|------------- |----- |-----------:|------:|
| CountManual  | 1    |   944.9 ns |  1.00 |
| CountJit     | 1    |   559.8 ns |  0.52 |
| CountPattern | 1    | 1,324.3 ns |  1.21 |
|              |      |            |       |
| CountManual  | 2    |   653.7 ns |  1.00 |
| CountJit     | 2    |   811.3 ns |  1.24 |
| CountPattern | 2    |   973.1 ns |  1.49 |
|              |      |            |       |
| CountManual  | 3    |   862.2 ns |  1.00 |
| CountJit     | 3    |   772.5 ns |  0.90 |
| CountPattern | 3    | 1,075.9 ns |  1.25 |

**Pull request merged**